### ISSUE TITLE
Choosing object or values shouldn't make a difference for calculations

### ISSUE THREAD

**Issue by fzumstein** [microsoft/python-in-excel#22]
Switching between the object and value representation should be purely a different view without having an impact on the results of further calculation steps. Right now, switching from object to values breaks this sheet:

Formulas:
<img width="973" alt="Screenshot 2023-09-08 at 8 30 02 AM" src="https://github.com/microsoft/python-in-excel/assets/2265504/95d5cf85-a2e2-4baf-914b-e3b766686ec5">

Object:
<img width="581" alt="Screenshot 2023-09-08 at 8 30 24 AM" src="https://github.com/microsoft/python-in-excel/assets/2265504/720cb974-d8c0-4e84-b99c-654c1cffb880">

Values:
<img width="518" alt="Screenshot 2023-09-08 at 8 30 52 AM" src="https://github.com/microsoft/python-in-excel/assets/2265504/190bca41-effc-4c65-818c-b4fff39f6142">

**Comment by jflam** [comment]
I don't think I understand why you expect that changing from by val to by ref would not break the sheet? In one case we are storing an object reference in a single cell; in the other case we are spilling the contents of a DataFrame across many cells.

What would you have it do instead?

**Comment by ncalm** [comment]
@fzumstein 
At the point the data are displayed, the content of D1 is no longer a data frame, it's a value: "a". 

If you put this in D1 instead, you can have both:

df = xl("A1:B3", headers=True)

This can display as values (spill to the sheet) and retain the ability for other cells to refer to df.

**Comment by fzumstein** [comment]
@jflam It's still a spreadsheet, isn't it? So if the formula doesn't change (and if it's a deterministic function), I don't expect it to have an influence on the calculations.

The (outdated?) docs https://support.microsoft.com/en-us/office/py-function-31d1c523-fb13-46ab-96a4-d1a90a9e512f seem to suggest that PY requires a `return_type` (0 - Excel value, 1 - Python object.): `=PY(python_code,return_type)`, which would improve the situation, but that's not how the formula works in my version of Excel (it doesn't have the `return_type`).

Objects in cells have been around for decades in about every single bank on this planet, and they always work the same: you can only return an object OR values to a cell. Consequently, if you want to view the values of the object, you'd do something like `=A1.values`, assuming that `A1` holds your object reference (that's how Excel data types like stocks work, anyway).

**Comment by b-gonzalez** [comment]
This seems like a trivial point. You could easily write a formula similar to the one you wrote on cell G1 like `xl("D1")` in Python Object mode. And from there just use the Excel Value option on that instead without breaking the references in G1 and H1.

**Comment by fzumstein** [comment]
@b-gonzalez That's close to my suggestion, only that is has to happen via formula (`=D1.values`), not via dropdown in the formula bar that can be changed by anybody without leaving a trace in the formula view.

**Comment by jflam** [comment]
I understand the confusion now. In reality the formula in the sheet is =PY(<code>, {0,1}). The UI in the formula editor toggles that last parameter, so it's also consistent with what our documentation says.

I think you perceive this as changing a formatting setting as opposed to changing the formula?

**Comment by fzumstein** [comment]
Yes, that's correct, that parameter is hidden in the formula, which is very confusing.

**Comment by keyur32** [comment]
Agree we should improve the documentation here. Tagging as a documentation issue.

**Comment by fzumstein** [comment]
I've just discovered that you can do `=A1.arrayPreview`, assuming that A1 has a DataFrame. That's basically what I'd want, but in the generic case, i.e., covering scalars as well. This makes it so much cleaner to understand that you're outside of the PY environment again, facing normal Excel cell values.

**Comment by fzumstein** [comment]
So having worked with PiE for a while, my conclusion is that referencing PY cells from other PY cells -- while possible and sort of a natural excel workflow -- should best be avoided. Instead, as @ncalm mentioned, just work with variables like you'd do in real Jupyter notebooks, too.
